tmt/base.py

```diff
@@ -1957,6 +1957,23 @@ def lint_fmf_remote_ids_valid(self) -> LinterReturn:
 
         yield LinterOutcome.SKIP, 'no remote fmf ids defined'
 
+    def lint_unique_names(self) -> LinterReturn:
+        """ P006: phases must have unique names """
+        passed = True
+        for step_name in self.step_names(enabled=True, disabled=True):
+            seen: List[str] = []
+            for phase in self._step_phase_nodes(step_name):
+                name = phase.get('name', None)
+                if name is not None:
+                    if name in seen:
+                        passed = False
+                        yield LinterOutcome.FAIL, \
+                            f'duplicate phase name "{name}" in step "{step_name}"'
+                    else:
+                        seen.append(name)
+        if passed:
+            yield LinterOutcome.PASS, 'phases have unique names'
```

Comment:
One more simplification maybe: `lint_uniq_names` and `_check_duplicate_names` do share some functionality, looking for duplicates. How about extracting that into a helper function? We already have a useful `tmt.utils.uniq`, I would add `tmt.utils.duplicates` to complement it:

```python
def duplicates(values: Iterable[Optional[T]]) -> Generator[T, None, None]:
    seen = collections.Counter(values)

    for value, count in seen.items():
        if value is None or count == 1:
            continue

        yield value

# And then in base.py...
def lint_unique_names(self) -> LinterReturn:
    """ P006: phases must have unique names """
    passed = True
    for step_name in self.step_names(enabled=True, disabled=True):
        for phase_name in duplicates(phase.get('name', None) for phase in self._step_phase_nodes(step_name)):
            passed = False
            yield LinterOutcome.FAIL, \
                f'duplicate phase name "{name}" in step "{step_name}"'

    if passed:
        yield LinterOutcome.PASS, 'phases have unique names'

# ... and steps/__init__.py:
def _check_duplicate_names(self, raw_data: List[_RawStepData]) -> None:
    """ Check for duplicate names in phases """

    for name in duplicates(raw_datum.get('name', None) for raw_datum in raw_data):
        raise tmt.utils.GeneralError(f"Duplicate phase name '{name}' in step '{self.name}'.")
```

We get a friend for `uniq`, a nicely isolated building block, a bit of a functional twist plus more readable methods.

Reply:
Thanks for your feedback. I extracted the shared functionality into a separate function, as you suggested. I also added a unit test for this new function.

Reply:
Nice!